### PR TITLE
Initial support for strings

### PR DESCRIPTION
**Pull request by isubasinghe**
This PR adds support for string slices (&str) and strings (String), however due to not being able to attach specifications to &str yet, new types (the StrSlice and String structs) are used as a temporary solution. 

StrSlice currently supports 4 operations:
  * StrSlice.len()
  * StrSlice.get_char(int) -> u8
  * StrSlice.is_ascii() -> bool
  * StrSlice.substring(int, int) -> StrSlice

### CODE REVIEW

**Comment by isubasinghe** [comment]
Note intent to squash and merge

File: source/vir/src/context.rs
Comment (isubasinghe) [inline source/vir/src/context.rs:78]: Is this unnecessary? a sha512 collision is extremely unlikely and even feels a bit silly to account for

File: source/vir/Cargo.toml
Comment (isubasinghe) [inline source/vir/Cargo.toml:14]: The hashing technique introduces two dependencies but the alternative of using a counter to assign an id to a string could be non deterministic. 

Perhaps the middle ground could be to use a 64 bit (or perhaps even 128 is fine) hash and then increment the integer until a collision is resolved.

File: source/vir/src/triggers_auto.rs
Comment (isubasinghe) [inline source/vir/src/triggers_auto.rs:340]: This doesn't seem to ever be triggered

Comment (isubasinghe) [reply source/vir/src/triggers_auto.rs:340]: I will hold off from opening a proper PR until this is resolved

File: source/rust_verify/tests/strings.rs
Comment (utaal) [inline source/rust_verify/tests/strings.rs:1]: Please add tests for:
* inequality and equality of string literals, both as they appear inline and as part of spec functions or e.g. postconditions of functions, i.e.
```
fn hello_str<'a>() -> (res: StrSlice<'a>)
  ensures res.substring(0, 5) === new_strlit("Hello")
```
and a function that uses the result of `hello_str`.

Comment (utaal) [reply source/rust_verify/tests/strings.rs:1]: Also, please add tests with multiple StrSlices and literals in the same program.

File: source/vir/src/prelude.rs
Comment (utaal) [inline source/vir/src/prelude.rs:216]: Add a header comment "String literals"

**Review by utaal — CHANGES_REQUESTED** [review]

File: source/vir/Cargo.toml
Comment (isubasinghe) [reply source/vir/Cargo.toml:14]: num-bigint is already used in the interpreter so this dependency is fine, I have decided to proceed with the approach as a result.

File: source/rust_verify/src/rust_to_vir_expr.rs
Comment (utaal) [inline source/rust_verify/src/rust_to_vir_expr.rs:993]: `strslice_get_char`?

File: source/rust_verify/src/rust_to_vir_expr.rs
Comment (utaal) [inline source/rust_verify/src/rust_to_vir_expr.rs:1008]: `strslice_len`?

File: source/rust_verify/src/rust_to_vir_expr.rs
Comment (utaal) [inline source/rust_verify/src/rust_to_vir_expr.rs:986]: Please briefly describe what's unexpected here, instead. "invalid parameter for strslice_get_char", for example.

File: source/rust_verify/src/rust_to_vir_func.rs
Comment (utaal) [inline source/rust_verify/src/rust_to_vir_func.rs:221]: Why this change?

File: source/vir/Cargo.toml
Comment (utaal) [reply source/vir/Cargo.toml:14]: 👍

File: source/vir/src/prune.rs
Comment (utaal) [inline source/vir/src/prune.rs:106]: Please add a note that this is temporary, until we support adding specification for `std`.

**Review by utaal — CHANGES_REQUESTED** [review]

File: source/rust_verify/src/rust_to_vir_func.rs
Comment (isubasinghe) [reply source/rust_verify/src/rust_to_vir_func.rs:221]: I think  had some trouble with my LSP and was trying to figure out what was causing {unknown} types to appear, this was an artifact of my LSP debugging that I forgot to remove

File: source/rust_verify/src/rust_to_vir_expr.rs
Comment (isubasinghe) [reply source/rust_verify/src/rust_to_vir_expr.rs]: Done

File: source/rust_verify/src/rust_to_vir_adts.rs
Comment (isubasinghe) [reply source/rust_verify/src/rust_to_vir_adts.rs]: not sure myself, reverted it

**Comment by utaal** [comment]
I'm going to approve this, but please wait for @Chris-Hawblitzel's review before merging.

**Review by utaal — APPROVED** [review]
`pervasive::String` is barebones, but we can expand it in a future PR.

**Comment by utaal** [comment]
Ah, one more thing: can you rename `.view()` to `.bytes()` (or similar), so we can keep `.view()` for when we introduce a character representation of the string?

**Review by Chris-Hawblitzel — APPROVED** [review]
I just have a couple of comments to address, then go ahead and merge.

**Pull request merged**
